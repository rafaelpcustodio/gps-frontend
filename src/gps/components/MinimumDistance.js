import PropTypes from 'prop-types'
import React from 'react'
import StyledMinimumDisance from './StyledMinimumDistance'


const MinimumDistance = props => {

    const { distance } = props
    
    return (
        <StyledMinimumDisance>
            { `Closest opened Service distance: ${distance}m` }
        </StyledMinimumDisance>
    )
}

MinimumDistance.propTypes = {
    distance: PropTypes.string
}

export default MinimumDistance