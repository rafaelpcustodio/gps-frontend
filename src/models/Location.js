class Location {

    constructor({ latitude, longitude }) {
        this._lat= latitude
        this._lng = longitude
    }

    get lat() {
        return this._lat
    }

    get lng() {
        return this._lng
    }

    get googleGpsLocation() {
        return {lat: this._lat, lng: this._lng}
    }
}

export default Location
