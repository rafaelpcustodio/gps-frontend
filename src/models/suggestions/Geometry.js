class Geometry {

    constructor({ location }) {
        this._location = location
    }

    get location() {
        return this._location
    }
}

export default Geometry