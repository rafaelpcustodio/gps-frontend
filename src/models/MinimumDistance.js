import Location from './Location'

class MinimumDistance {

    constructor({ location, minimumDistance }) {
        this._location = new Location(location)
        this._distance = minimumDistance
    }

    get distance() {
        return this._distance
    }

    get location() {
        return this._location
    }

    get toGoogleMaps() {
        return this.location.toGoogleMaps
    }
}

export default MinimumDistance