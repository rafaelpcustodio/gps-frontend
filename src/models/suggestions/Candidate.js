import Geometry from './Geometry'
import OpeningHours from './OpeningHours'
import { getIcon } from '../../utils/functions'

class Candidate {

    constructor({ geometry, name, opening_hours = '', rating, types, business_status }) {
        this._geometry = new Geometry(geometry)
        this._name = name
        this._rating = rating
        this._opening_hours = new OpeningHours(opening_hours)
        this._types = types
    }

    get opening_hours() {
        return this._opening_hours
    }

    get geometry() {
        return this._geometry
    }

    get name() {
        return this._name
    }

    get rating() {
        return this._rating
    }

    get types() {
        return this._types
    }

    getIcon() {
        return getIcon(this.types)
    }
    
}

export default Candidate