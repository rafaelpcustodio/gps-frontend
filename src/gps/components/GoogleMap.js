import PropTypes from 'prop-types'
import React from 'react'
import { Map, GoogleApiWrapper, Marker, Polyline } from 'google-maps-react'
import truckMarker from '../../_assets/img/icn-current-location.png'
import pathMarker from '../../_assets/img/icn-path.png'
import firstLocationMarker from '../../_assets/img/icn-first-location.png'

import Location from '../../models/Location'
import Candidate from '../../models/suggestions/Candidate'

import { isNotEmpty } from '../../utils/functions'

import { colors } from '../../helpers/colors'

const GoogleMap = props => {

    const {
        minimumDistance,
        locationsList,
        suggestionsList
    } = props

    const mapStyles = {
        width: '100%',
        height: '100%',
    };
    
    const DEFAULT_CENTER = {
        lat: 41.40298,
        lng: 2.17398
    }

    const arrow = {
        path: window.google.maps.SymbolPath.FORWARD_CLOSED_ARROW
    }
    
    return (
        <Map
            google={props.google}
            disableDefaultUI={true}
            zoom={16}
            style={mapStyles}
            initialCenter={DEFAULT_CENTER}
            >
                {
                    locationsList[0] &&
                    ( 
                        <Marker 
                            icon={truckMarker}
                            position={locationsList[0]}
                        />
                    ) 
                }
                {
                    isNotEmpty(minimumDistance) && isNotEmpty(locationsList[0]) && (
                        <Polyline 
                            geodesic={true}
                            strokeColor={colors.lightGreen}
                            strokeOpacity={1.0}
                            icons={[{icon: arrow, offset: '100%'}]}
                            strokeWeight={2}
                            path={[locationsList[0].toGoogleMaps, minimumDistance.toGoogleMaps]}
                        /> 
                    )
                }
                {
                    locationsList && locationsList.map((location, index) => {
                        if(index === locationsList.length - 1) {
                            return(<Marker 
                                icon={firstLocationMarker}
                                key={index}
                                position={location}
                            />)
                        } else if(index !== 0) {
                            return(<Marker 
                                icon={pathMarker}
                                key={index}
                                position={location}
                            />)  
                        }
                    })
                }
                {
                    suggestionsList && suggestionsList.map((suggestion, index) => 
                        <Marker
                            icon={suggestion.getIcon()}
                            key={index}
                            position={suggestion.toGoogleMaps}
                        />
                    )
                }
            </Map>
    );
}

GoogleMap.defaultProps = {
    minimumDistance: '',
    suggestionsList: [],
    locationsList: []
}

GoogleMap.protoTypes = {
    minimumDistance: PropTypes.string,
    suggestionsList: PropTypes.arrayOf(PropTypes.objectOf(Candidate)),
    locationsList: PropTypes.arrayOf(PropTypes.objectOf(Location))
}

export default GoogleApiWrapper({
    apiKey: process.env.REACT_APP_GOOGLE_API_KEY
  })(GoogleMap);