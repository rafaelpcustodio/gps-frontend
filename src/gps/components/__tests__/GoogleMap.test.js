import configureMockStore from 'redux-mock-store'
import { Map, Record } from 'immutable'
import React from 'react'
import renderer from 'react-test-renderer'

import { Provider } from 'react-redux'

import MinimumDistance from '../../../models/MinimumDistance'
import Candidate from '../../../models/suggestions/Candidate'

import GoogleMap from '../GoogleMap'

import Location from '../../../models/Location'
import Geometry from '../../../models/suggestions/Geometry'

describe('Tests for GoogleMap component', () => {
    let minimumDistance = {}
    let location = {}
    let suggestionsList = {}
    let locationsList = []
    let store = {}
    let mockStore = {}

    beforeEach(() => {
        minimumDistance = new MinimumDistance({location:{ latitude: 1.453, longitude: 1.453 }, minimumDistance: 400})
        suggestionsList = new Candidate({ geometry: new Geometry(location), name: 'Hotel', rating: 1, types: ['a','b'] })
        locationsList = [ new Location({ lat: 1.453, lng: 1.453 }) ]
    })

    jest.mock('../_assets/icn-restaurant.png')
    jest.mock('../_assets/icn-hotel.png')
    jest.mock('../_assets/icn-hotel-restaurant.png')
    jest.mock('../_assets/icn-gas-station.png')
    jest.mock('../_assets/icn-gas-station.png')
    jest.mock('@fortawesome/react-fontawesome', () => ({
        FontAwesomeIcon: ''
    }))

    it('Should render without crash', () => {
        mockStore = configureMockStore()
        store = mockStore(Record({
            gpsReducers: Map({
                licensePlateSelected: 'ABC1234',
                minimumDistance: {},
                noLocations: false,
                poiTypeList: [],
                poiTypeSelected: '',
                radiusList: [],
                radiusSelected: '',
                suggestionsList: suggestionsList,
                locationsList: locationsList
            })
        })())

        const tree = renderer.create(
            <Provider store={ store }>
                <GoogleMap
                    requestGetSuggestionsAction={jest.fn()}
                    minimumDistance={minimumDistance}s
                    locationsList={locationsList}
                    suggestionsList={suggestionsList}
                />
            </Provider>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })
})
