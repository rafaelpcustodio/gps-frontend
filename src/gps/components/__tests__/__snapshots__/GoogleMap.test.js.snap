// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for GoogleMap component Should render without crash 1`] = `
<div>
  Loading...
</div>
`;

exports[`Tests for Gps component Should render without crash 1`] = `
Array [
  <div
    className="SearchBarContainer-sc-1byzrjc-0 janIPg"
  >
    <input
      className="LicensePlate__StyledLicensePlateText-pf7qoz-0 ksfRmt"
      maxLength="7"
      onChange={[Function]}
      placeholder="License plate (ex: ABC1234)"
      value=""
    />
    <div
      className="StyledSelect-sc-17xe3wx-0 cnxjXs"
      disabled={true}
    >
      <div
        className="StyledSelectedContainer__StyledSelectContainer-sc-2m0ndn-0 fwyTFq"
      >
        <span
          className="SelectPoiTypeContainer__SelectText-sc-19t5xdi-0 iSHPxC"
          disabled={true}
          onClick={[Function]}
        >
          Select POI type
        </span>
        <button
          className="StyledDropdown-sc-1d17mlt-0 crWdhk"
        />
      </div>
      <div
        className="StyledSelectList-vv0ppr-0 djStfM"
      >
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          Restaurants
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          Hotels
        </li>
      </div>
    </div>
    <div
      className="StyledSelect-sc-17xe3wx-0 cnxjXs"
      disabled={true}
    >
      <div
        className="StyledSelectedContainer__StyledSelectContainer-sc-2m0ndn-0 fwyTFq"
      >
        <span
          className="SelectRadiusContainer__SelectText-sz6f5m-0 heeAuE"
          disabled={true}
          onClick={[Function]}
        >
          Select radius
        </span>
        <button
          className="StyledDropdown-sc-1d17mlt-0 crWdhk"
        />
      </div>
      <div
        className="StyledSelectList-vv0ppr-0 djStfM"
      >
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          1000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          2000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          4000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          4500
        </li>
      </div>
    </div>
    <button
      className="StyledButton-sc-1z0vb0n-0 ejBCHp"
      onClick={[Function]}
    >
      <div
        className="Button__StyledText-ko51rl-0 eowQDx"
      >
        Apply
      </div>
    </button>
  </div>,
  <div
    className="StyledMinimumDistance__StyledMinimumDisance-vqza48-0 kYkUOE"
  >
    Closest opened Service: 400m
  </div>,
  <div>
    Loading...
  </div>,
]
`;

exports[`Tests for Gps component Should render without crash 2`] = `
Array [
  <div
    className="SearchBarContainer-sc-1byzrjc-0 janIPg"
  >
    <input
      className="LicensePlate__StyledLicensePlateText-pf7qoz-0 ksfRmt"
      maxLength="7"
      onChange={[Function]}
      placeholder="License plate (ex: ABC1234)"
      value=""
    />
    <div
      className="StyledSelect-sc-17xe3wx-0 cnxjXs"
      disabled={true}
    >
      <div
        className="StyledSelectedContainer__StyledSelectContainer-sc-2m0ndn-0 fwyTFq"
      >
        <span
          className="SelectPoiTypeContainer__SelectText-sc-19t5xdi-0 iSHPxC"
          disabled={true}
          onClick={[Function]}
        >
          Select POI type
        </span>
        <button
          className="StyledDropdown-sc-1d17mlt-0 crWdhk"
        >
          <
            className=""
            color="#aaaaaa"
            data-cy="icon"
            icon="caret-down"
            size="lg"
          />
        </button>
      </div>
      <div
        className="StyledSelectList-vv0ppr-0 djStfM"
      >
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          Restaurants
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          Hotels
        </li>
      </div>
    </div>
    <div
      className="StyledSelect-sc-17xe3wx-0 cnxjXs"
      disabled={true}
    >
      <div
        className="StyledSelectedContainer__StyledSelectContainer-sc-2m0ndn-0 fwyTFq"
      >
        <span
          className="SelectRadiusContainer__SelectText-sz6f5m-0 heeAuE"
          disabled={true}
          onClick={[Function]}
        >
          Select radius
        </span>
        <button
          className="StyledDropdown-sc-1d17mlt-0 crWdhk"
        >
          <
            className=""
            color="#aaaaaa"
            data-cy="icon"
            icon="caret-down"
            size="lg"
          />
        </button>
      </div>
      <div
        className="StyledSelectList-vv0ppr-0 djStfM"
      >
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          1000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          2000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          4000
        </li>
        <li
          className="SelectItem__StyledSelectItem-c7nnhy-0 lkKUay"
          onClick={[Function]}
        >
          4500
        </li>
      </div>
    </div>
    <button
      className="StyledButton-sc-1z0vb0n-0 ejBCHp"
      onClick={[Function]}
    >
      <div
        className="Button__StyledText-ko51rl-0 eowQDx"
      >
        Apply
      </div>
    </button>
  </div>,
  <div
    className="StyledMinimumDistance__StyledMinimumDisance-vqza48-0 kYkUOE"
  >
    Closest opened Service: 400m
  </div>,
  <div>
    Loading...
  </div>,
]
`;
